File: python/solid/liskov_substitution.py
class Rectangle:
    def __init__(self, width, height):
        self._width = width
        self._height = height

    def get_width(self):
        return self._width

    def set_width(self, width):
        self._width = width

    def get_height(self):
        return self._height

    def set_height(self, height):
        self._height = height


"""
    Here the Square class inheritance from Rectangle but should override the setter methods to kkep it square
    and this way we will get unexpected result
    this breaks the liskov rule
"""


class Square(Rectangle):
    def __init__(self, size):
        super().__init__(size, size)

    def set_width(self, width):
        self._width = width
        self._height = width

    def set_height(self, height):
        self._width = height
        self._height = height


def use_it(rectangle: Rectangle):
    width = rectangle.get_width()
    rectangle.set_height(10)

    expected_area = 10 * width
    actual_area = rectangle.get_width() * rectangle.get_height()

    print(f"Expected: {expected_area}, actual: {actual_area}")


# Got unexpected for the square
rect = Rectangle(2, 3)
use_it(rect)  # Expected area: 20, Actual area: 20

sq = Square(5)
use_it(sq)  # Expected area: 50, Actual area: 100

# we should separate the square from the Rectangle for a solution to this

File: golang/solid/liskov_substitution.go
package solid

import "fmt"

type Sized interface {
	GetWidth() int
	SetWidth(width int)
	GetHeight() int
	SetHeight(height int)
}

type Rectangle struct {
	width, height int
}

func (r *Rectangle) GetWidth() int        { return r.width }
func (r *Rectangle) SetWidth(width int)   { r.width = width }
func (r *Rectangle) GetHeight() int       { return r.height }
func (r *Rectangle) SetHeight(height int) { r.height = height }

type Square struct {
	Rectangle
}

func NewSquare(size int) *Square {
	square := Square{}
	square.height = size
	square.width = size
	return &square
}

// we should set both high and width to keep this square

func (s *Square) SetWidth(width int) {
	s.width = width
	s.height = width
}

func (s *Square) SetHeight(height int) {
	s.width = height
	s.height = height
}

// here in UseIt we will have wrong result with Square and we already violates the liskov here

func UseIt(sized Sized) {
	width := sized.GetWidth()
	sized.SetHeight(10)
	expectedArea := 10 * width
	actualArea := sized.GetHeight() * sized.GetWidth()

	fmt.Println("expected: ", expectedArea, "actual: ", actualArea)
}

// solution is to do like this

type Square2 struct {
	size int // width, height
}

func (s *Square2) Rectangle() Rectangle {
	return Rectangle{s.size, s.size}
}

// having rectangle and square together with the same interface
//will break the liskov substitution and should be separated
